### PR TITLE
add a method to stop the mixer Store explicitly

### PR DESCRIPTION
**Pull request by jmuk**
Previously it accepted context.Context for the validity of the
store, but that does not fit well with the mixer's calling
structure. Also context's document says a struct shouldn't store
the context as a field.

This PR adds `Stop` method to do this.

### CODE REVIEW

File: mixer/pkg/runtime2/runtime.go [hunk 1]
```diff
@@ -141,6 +141,9 @@ func (c *Runtime) StopListening() {
 	if c.shutdown != nil {
 		c.shutdown <- struct{}{}
 		c.shutdown = nil
+		if err := c.store.Close(); err != nil {
+			log.Errorf("Failed on close: %v", err)
```
Comment (ozevren) [inline mixer/pkg/runtime2/runtime.go]: Can you improve the error message (i.e.  "store.Close failed during runtime shutdown: %v")

File: mixer/pkg/runtime2/runtime.go [hunk 2]
```diff
@@ -141,6 +141,9 @@ func (c *Runtime) StopListening() {
 	if c.shutdown != nil {
 		c.shutdown <- struct{}{}
 		c.shutdown = nil
+		if err := c.store.Close(); err != nil {
```
Comment (ozevren) [inline mixer/pkg/runtime2/runtime.go]: Should the close be before c.waitQuiesceListening, or after? 
The background go-routine will be listening to both the c.shutdown channel, as well as the channel returned by the store. When the store is closed here, what will happen to the channel that is returned by the store? (Or

Comment (jmuk) [reply mixer/pkg/runtime2/runtime.go]: Ooh, nice catch!  The watch channel shall be closed on store's `Close()`, that means the ordering is incorrect. `watchChanges` might not finish at all. Changed the order, waitQuiesceListening first, and then store.Close. Thank you!

File: mixer/pkg/runtime2/runtime.go (hunk 1 above)
Comment (jmuk) [reply mixer/pkg/runtime2/runtime.go]: Done

**Comment by ldemailly** [comment]
@ozevren is this good to go now ? if so please / l g t m

File: mixer/pkg/config/crd/store.go [hunk 3]
```diff
@@ -111,13 +110,18 @@ type Store struct {
 var _ store.Backend = &Store{}
 var _ probe.SupportsProbe = &Store{}
 
+// Close implements io.Closer
```
Comment (ozevren) [inline mixer/pkg/config/crd/store.go]: Do we have to implement io.Closer? I don't think we're using this interface to reference the Store.

It looks to me, if we simply have a Close() method that returns nothing, we can simplify some of the client code paths.

File: mixer/pkg/config/crd/store.go [hunk 4]
```diff
@@ -111,13 +110,18 @@ type Store struct {
...
+func (s *Store) Close() error {
+	close(s.donec)
+	return nil
+}
+
 // checkAndCreateCaches checks the presence of custom resource definitions through the discovery API,
 // and then create caches through lwBUilder which is in kinds. It retries within the timeout duration.
```
Comment (ozevren) [inline mixer/pkg/config/crd/store.go]: Update comments? There is no timeout anymore.

Comment (jmuk) [reply mixer/pkg/config/crd/store.go]: done

**Comment by jmuk** [comment]
PTAL; also updated the PR title / comment.

Note that the failures seem to be unrelated side effects of #3529, nothing by this PR.

**Comment by yutongz** [comment]
/test e2e-simple
/test e2e-bookInfo

We renamed e2e-smoke to e2e-bookInfo and added e2e-simple
ref: https://github.com/istio/istio/pull/3529

**Pull request merged**